src/TodoContext.tsx

```diff
@@ -0,0 +1,319 @@
+import React, { useState, useEffect } from 'react';
+import { Todo } from './types/Todo';
+import {
+  addTodo, completeTodo, deleteTodo, getTodos,
+} from './api/todos';
+
+export enum Key {
+  Enter = 'Enter',
+  Escape = 'Escape',
+}
+
+type PartialTodo = Omit<Todo, 'id'>;
+
+const USER_ID = 11880;
+
+export enum FilterBy {
+  All = 'all',
+  Active = 'active',
+  Completed = 'completed',
+}
+
+const filterTodos = (todos: Todo[], filter: FilterBy): Todo[] => {
+  switch (filter) {
+    case FilterBy.All:
+      return todos;
+    case FilterBy.Active:
+      return todos.filter(todo => !todo.completed);
+    case FilterBy.Completed:
+      return todos.filter(todo => todo.completed);
+    default:
+      return todos;
+  }
+};
+
+type Props = {
+  children: React.ReactNode,
+};
+
+interface Context {
+  allTodos: Todo[],
+  setError: React.Dispatch<React.SetStateAction<string>>
+  filteredTodos: Todo[],
+  error: string,
+  completedTodos: Todo[],
+  activeTodos: Todo[],
+  filterBy: FilterBy,
+  newTodoTitle: string,
+  tempTodo: Todo | null,
+  isBlured: number | null,
+  allBlured: boolean,
+  isRendering: boolean,
+  onSubmit: (e: React.FormEvent) => void
+  onDelete: (todoId: number) => void
+  clearCompleted: () => void
+  onCompleteChange: (e: React.ChangeEvent<HTMLInputElement>,
+    postId: number) => void
+  setAllCompleted: () => void
+  handleFilterClick: (filterType: FilterBy) => (
+    event: React.MouseEvent,
+  ) => void
+  setIsBlured: React.Dispatch<React.SetStateAction<number | null>>
+  handleTitleChange: (e: React.ChangeEvent<HTMLInputElement>) => void,
+  setAllTodos: React.Dispatch<React.SetStateAction<Todo[]>>,
+}
+
+export const TodosContext = React.createContext<Context>({
+  allTodos: [],
+  filteredTodos: [],
+  error: '',
+  filterBy: FilterBy.All,
+  newTodoTitle: '',
+  tempTodo: null,
+  isBlured: null,
+  allBlured: false,
+  isRendering: false,
+  handleTitleChange: () => {},
+  onSubmit: () => {},
+  onDelete: () => {},
+  clearCompleted: () => {},
+  onCompleteChange: () => {},
+  setAllCompleted: () => {},
+  handleFilterClick: () => () => {},
+  completedTodos: [],
+  activeTodos: [],
+  setError: () => {},
+  setAllTodos: () => {},
+  setIsBlured: () => {},
+});
+
+export const TodosProvider: React.FC<Props> = ({ children }) => {
+  const [allTodos, setAllTodos] = useState<Todo[]>([]);
+  const [filteredTodos, setFilteredTodos] = useState<Todo[]>([]);
+  const [error, setError] = useState('');
+  const [filterBy, setFilterBy] = useState<FilterBy>(FilterBy.All);
+  const [newTodoTitle, setNewTodoTitle] = useState('');
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+  const [isBlured, setIsBlured] = useState<number | null>(null);
+  const [allBlured, setAllBlured] = useState(false);
+  const [isRendering, setIsRendering] = useState(false);
```

Comment:
looks like instead these three states (isBlured, allBlured, isRendering) can be one state which will be array of todos which in process (updating, removing)
set id of processing todos in state and remove them from state when receive response from server.
base on this state show or not loader.